### PR TITLE
0.4.28 Add unit tests, coverage, and fix minor issues.

### PR DESCRIPTION
**Pull request by erlendcognite**

### CODE REVIEW

**Comment by psalaberria002** [comment]
Good start. Try to fix conflicts, and squash before merging this (you can do squash and merge in Github).

Change the PR title to something other than `Tests`

File: Pipfile [hunk 1]
```diff
@@ -1,19 +1,27 @@
...
 [packages]
+
 twine = "*"
 Sphinx = "*"
 "yolk3k" = "*"
 pylint = "*"
 pandas = "*"
 requests = "*"
 protobuf = "*"
+pytest = "*"
```
Comment (psalaberria002) [inline Pipfile]: Could pytest and pylint be part of dev-packages also?

File: tests/test_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        mask = df_nan_col_removed[1]
+        correct_mask = not (mask[0] + mask[2]).any() and (mask[1] + mask[3]).all()
+        assert correct_mask
+
+    def test_correct_column_removed(self, df_nan_col_removed):
+        assert 0 not in df_nan_col_removed[0].columns
+
+
+class TestRemoveZeroVarColumns():
+    @pytest.fixture
```
Comment (psalaberria002) [inline tests/test_preprocessing.py]: Is this necessary? You are not adding any fixture

File: tests/test_tag_matching.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+def mocked_post_request():
+    return {'cucoo': 'haha'}
+
+
+@patch('requests.post')
+@pytest.fixture(scope='module')
+def tagmatching_result(mock_post):
+    from cognite.tagmatching import tag_matching
+    mock_post.return_value.status_code = 200
```
Comment (psalaberria002) [inline tests/test_tag_matching.py]: Only mocking status code and return value is risky if the functions try to do something more fancy than only reading those values. Try to use MockReturnValue in conftest.

File: tests/test_tag_matching.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    mock_post.return_value.json.return_value = {'data': {'items': [{'matches': [
+        {'platform': 'a_platform', 'score': 0, 'tagId': 'a_match'},
+        {'platform': 'a_platform', 'score': 0, 'tagId': 'a_match1'},
+        {'platform': 'a_platform', 'score': 0, 'tagId': 'a_match2'}], 'tagId': 'a_tag'}]}}
+    return tag_matching(tag_ids=['a_tag'])
+
+
+def test_object(tagmatching_result):
```
Comment (psalaberria002) [inline tests/test_tag_matching.py]: How does this work? Doesn't it shadow tagmatching_result ? I tried to run it locally, and it is required... Is this some sort of dependency injection?

File: tests/test_timeseries.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+from datetime import datetime
+from unittest.mock import patch
+
+import pytest
+
+from cognite.timeseries import get_datapoints
+
+
+@patch('requests.get')
+@pytest.fixture(scope='module', params=[
```
Comment (psalaberria002) [inline tests/test_timeseries.py]: nice

File: tests/test_timeseries.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    {'start': 0, 'end': 1},
+    {'start': '1w-ago', 'end': '1d-ago'},
+    {'start': datetime(2018, 3, 26), 'end': datetime(2018, 3, 27)}])
+def get_datapoints_response_obj(mock_get, request):
+    mock_get.return_value.status_code = 200
```
Comment (psalaberria002) [inline tests/test_timeseries.py]: Same as for the previous one. I would prefer a more proper mock object such as MockReturnValue in conftest

File: tests/test_timeseries.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                     {'timestamp': 4, 'value': 50},
+                 ]
+                 }
+            ]
+        }
+    }
+
+    return get_datapoints(tag_id='SKAP_18PI2317/Y/10sSAMP', start=request.param['start'], end=request.param['end'],
+                          protobuf=False)
+
```
Comment (psalaberria002) [inline tests/test_timeseries.py]: Add a test checking the content. Length of items, or datapoints for tagid

File: tests/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+@pytest.mark.usefixtures('configure_test_session')
+def url():
+    api_key, project = config.get_config_variables(None, None)
+    return config.get_base_url() + '/projects/{}/assets'.format(project)
+
+
+class TestRequests:
+    @mock.patch('cognite._utils.requests.get')
+    def test_get_request(self, mock_request):
+        mock_request.return_value = MockReturnValue(json_data=RESPONSE)
```
Comment (psalaberria002) [inline tests/test_utils.py]: This looks better ;)

File: tests/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        mock_request.side_effect = check_gzip_disabled_and_return_mock
+
+        response = utils.post_request(
+            url, RESPONSE,
+            headers={},
+            use_gzip=False)
+        assert response.status_code == 200
+
+
+        # TODO: uncomment these tests when merged into master, put_request does not exist in this branch.
```
Comment (psalaberria002) [inline tests/test_utils.py]: Are you sure? I think they do exist. Uncomment this section and give it a try

File: tests/test_tag_matching.py (hunk 4 above)
Comment (erlendcognite) [reply tests/test_tag_matching.py]: yes, that is how you inject fixtures into tests.

File: Pipfile (hunk 1 above)
Comment (erlendcognite) [reply Pipfile]: yes, twine and yolk aswell i guess

**Pull request merged**